lib/features/build_server/build_provider.dart

```diff
@@ -0,0 +1,265 @@
+  /// the project with the id [id].
+  bool _isFirstInQueue(ServerSession session, String id) {
+    if (_projectCapacityQueues[id]?.isNotEmpty == true) {
+      return identical(_projectCapacityQueues[id]?.first, session);
+    }
+
+    return false;
+  }
+
+  /// Activates the first session in the queue for the project with the id [id].
+  ///
+  /// If [removing] is not null, the session is removed from the queue.
+  Future<void> _activateNextSession(String id, {ServerSession? removing}) async {
+    if (removing != null) {
+      _projectCapacityQueues[id]?.remove(removing);
+    }
+
+    if (_projectCapacityQueues[id]?.isNotEmpty == true) {
+      _activateSession(_projectCapacityQueues[id]!.first, id);
+    }
+  }
+
+  /// Activates the [session] for the project with the id [id].
+  Future<void> _activateSession(ServerSession session, String id) async {
+    _interfaces[id]?.onStateUpdate = session.send;
+    session.send(isActiveSessionResponse);
+  }
+
+  /// Returns the interface with the given [id] or throws an exception.
+  MainInterface _getInterface(String id) {
+    if (_interfaces.containsKey(id)) {
+      return _interfaces[id]!;
+    }
+
+    throw StateError('Interface with id $id has not been created yet.');
+  }
+
+  /// Throws an error if the [session] has no id or returns it.
+  String _ensureId(ServerSession session) {
+    final dynamic id = session.storage['id'];
+
+    if (id is String) {
+      return id;
+    }
+
+    throw StateError('Session has no project id.');
+  }
+
+  /// Handles a new project initialization request.
+  Future<void> _handleProjectInit(ServerSession session, ProjectInitRequest message) async {
+    if (_interfaces.containsKey(message.id)) {
+      return;
+    }
+
+    _interfaces[message.id] = await MainInterface.launch();
+    _availableProjects[message.id] = message.name;
+    session.storage['id'] = message.id;
+    _interfaces[message.id]?.onStateUpdate = session.send;
+    _projectCapacityQueues[message.id] = Queue<ServerSession>();
+
+    session.send(
+      ProcessCompletionMessage(
+        process: CompletableProcess.projectInit,
+        description: '',
+      ),
+    );
+  }
+
+  /// Handles a new project update request.
+  Future<void> _handleProjectUpdate(
+    ServerSession session,
+    ProjectUpdateRequest message,
+    String id,
+  ) async {
+    final ProjectCache cache;
+
+    if (!_cache.has(id)) {
+      cache = await _cache.create(id);
+    } else {
+      cache = await _cache.get(id);
+    }
+
+    _getInterface(id).init(message.files, message.config, cache);
+
+    session.send(
+      ProjectUpdateResponse(
+        files: await _getInterface(id).sync(
+          message.files,
+          message.config,
+        ),
+      ),
+    );
+  }
+
+  /// Handles a new file update request.
+  Future<void> _handleFileUpdate(
+    ServerSession session,
+    FileUpdate message,
+    String id,
+  ) async {
+    final file = SourceFile(
```

Comment:
Why not use `message.toSourceFile()`?

Reply:
This change would also bring the test coverage of _core/networking/protocol/project_update.dart_ to 100%.